### ISSUE TITLE
max_zoom and default tiles

### ISSUE THREAD

**Issue by hupe13** [bozdoz/wp-plugin-leaflet-map#232]
Hi Bozdoz,

Due to a bug in my plugin, I took one more look at max_zoom.  I would like to change 
https://github.com/bozdoz/wp-plugin-leaflet-map/blob/b6bafe8f4fd1373776201768f5a179d55fcdc845/shortcodes/class.map-shortcode.php#L81-L82
to
```php
        $atts['max_zoom'] = empty($max_zoom) ?
            $settings->get('default_max_zoom') : 
            ($max_zoom <= $settings->get('default_max_zoom') ? $max_zoom : $settings->get('default_max_zoom'));
```

Default max zoom should be the max zoom of default tiles or smaller.

I make a pull request.

**Comment by bozdoz** [comment]
I don't understand. If the user passes `max-zoom`, why would we override that with the default max zoom?

**Comment by hupe13** [comment]
Some tiles have a max zoom and the max zoom from map must not be greater as the max zoom from tiles. If the max zoom from map is greater as max zoom from tiles you get grey areas.

**Comment by bozdoz** [comment]
Just seems like a difficult problem to debug if someone *wants* to override the default but can't

**Comment by hupe13** [comment]
Or we need to have two options max_zoom, one for map and one for tiles.
And the same for min_zoom.
https://leafletjs.com/reference.html#map-minzoom
https://leafletjs.com/reference.html#map-maxzoom

**Comment by hupe13** [comment]
New approach:

https://github.com/bozdoz/wp-plugin-leaflet-map/blob/b6bafe8f4fd1373776201768f5a179d55fcdc845/shortcodes/class.map-shortcode.php#L193-L201

We need some more options. There is a repository [leaflet-providers](https://github.com/leaflet-extras/leaflet-providers) and here are the options for many [tileservers](https://github.com/leaflet-extras/leaflet-providers/blob/master/leaflet-providers.js). Is there a convenient way to consider many different options without defining a setting for each possible one? For now you have defined some options for MapQuest.

To avoid grey areas there is an option [maxNativeZoom](https://leafletjs.com/reference.html#gridlayer-maxnativezoom).

**Comment by bozdoz** [comment]
I suppose min and max zoom overlap too much with tilelayer min and max; why don't we just add `tileMaxZoom`?

**Comment by hupe13** [comment]
This I have done now. I prepared a [pull request](https://github.com/bozdoz/wp-plugin-leaflet-map/compare/master...hupe13:wp-plugin-leaflet-map:master). Sorry about the removed spaces, this makes my editor automatically.

class.leaflet-map.php:
-  leafletjs version 1.9.4

class.plugin-settings.php
- changed default max_zoom to 18, because https://leafletjs.com/reference.html#map-maxzoom and https://leafletjs.com/reference.html#tilelayer-maxzoom
- removed subdomains from default map_tile_url
- new option map_tile_maxzoom

readme.txt
- WordPress 6.3

shortcodes/class.map-shortcode.php
- changed getting map maxZoom, if empty
- tile maxZoom

templates/settings.php:
- added _e(' - Please change JavaScript-URL and CSS-URL.', 'leaflet-map');

What do you mean?

**Comment by bozdoz** [comment]
Thanks! I'll take a look next week.

**Comment by bozdoz** [comment]
Why remove subdomains?

**Comment by hupe13** [comment]
I read this in [german OSM forum](https://community.openstreetmap.org/t/a-b-c-tile-openstreetmap-de-subdomains-von-tile-openstreetmap-de-werden-aufgehoben/100830). There is a link to a [issue](https://github.com/openstreetmap/operations/issues/737). And in leaflet-providers.js are the [tiles without subdomain](https://github.com/leaflet-extras/leaflet-providers/blob/176e603a7eeb155b754541e96fd9ef53c0241e6f/leaflet-providers.js#L80C30-L80C30).

**Comment by bozdoz** [comment]
Great find. Makes sense

**Comment by bozdoz** [comment]
Could you make a pull request of that branch?

**Comment by hupe13** [comment]
max_zoom and map_tile_maxzoom do not work yet together, and it takes time to test. If I write
`[leaflet-map zoom=18 max_zoom=23 map_tile_maxzoom="20" tileurl=https://....]`
I get grey tiles if I zoom to a zoomlevel greater than map_tile_maxzoom.

**Comment by bozdoz** [comment]
Right but what would you expect given a shortcode like that? Are you suggesting that map max zoom should be restricted to map tile max zoom no matter what?

**Comment by hupe13** [comment]
I made a [commit](https://github.com/hupe13/wp-plugin-leaflet-map/commit/b1aad92631dd24f128f01120729022c4c6b5ccb2). Now it works as I expect. But it does not work, if detect-retina is true and it is a retina display. Then the maxZoom should be one step lower.

**Comment by hupe13** [comment]
I think, I found the solution, it works with Retina also now.

The differences to your reposiitory are:

- changed default max_zoom to 18, because https://leafletjs.com/reference.html#map-maxzoom and https://leafletjs.com/reference.html#tilelayer-maxzoom
- removed subdomains from default map_tile_url
- new option map_tile_maxzoom
- new option map_tile_minzoom
- no grey tiles anymore

Maybe we need to change "Info: your leaflet version may be out-of-sync with the latest default version: 1.9.4".  Many don't know what to do with it. But you do it, your english is better than mine.

What do you mean?

**Comment by bozdoz** [comment]
Yeah I looked into changing it. Maybe we could output the js and css urls we expect for latest.

**Comment by hupe13** [comment]
Closed because outdated.
